### PR TITLE
fix(@angular/cli): Disable comparisons feature in uglify compression in prod

### PR DESCRIPTION
**Pull request by Wykks**
See #5804

### CODE REVIEW

**Comment by Wykks** [comment]
I signed it!

**Comment by Wykks** [comment]
I signed it!

**Comment by Wykks** [comment]
Some size tests : 
A small app (which use mapbox-gl) with `comparisons: false`
```
chunk {0} polyfills.d8d3d78a4deb2ab66856.bundle.js (polyfills) 66.2 kB {4} [initial] [rendered]
chunk {1} main.e0d1f68923b143f095a6.bundle.js (main) 40.9 kB {3} [initial] [rendered]
chunk {2} styles.cbb57e89e9ac6ad6f3fb.bundle.css (styles) 78.1 kB {4} [initial] [rendered]
chunk {3} vendor.4ad7ad49f3f7f165fff5.bundle.js (vendor) 1.08 MB [initial] [rendered]
chunk {4} inline.94dcaa1fbb222b2f330f.bundle.js (inline) 1.45 kB [entry] [rendered]
```
Without `comparisons: false`
```
chunk {0} polyfills.d8d3d78a4deb2ab66856.bundle.js (polyfills) 66.1 kB {4} [initial] [rendered]
chunk {1} main.e0d1f68923b143f095a6.bundle.js (main) 40.9 kB {3} [initial] [rendered]
chunk {2} styles.cbb57e89e9ac6ad6f3fb.bundle.css (styles) 78.1 kB {4} [initial] [rendered]
chunk {3} vendor.4ad7ad49f3f7f165fff5.bundle.js (vendor) 1.08 MB [initial] [rendered]
chunk {4} inline.94dcaa1fbb222b2f330f.bundle.js (inline) 1.45 kB [entry] [rendered]
```

**Comment by filipesilva** [comment]
@Wykks can you check out my alternative solution in https://github.com/angular/angular-cli/issues/5804#issuecomment-336073573 please? I prefer that one if it also works.

**Comment by khanhvu161188** [comment]
the PR worked well. Hope it will be merged soon

**Comment by Wykks** [comment]
Tests with a slightly bigger project (angular/cli 1.5 & angular 5): 
Without `comparisons: false`, prod buid (mapbox-gl doesn't work):
```
chunk {0} 0.7120f44b0b35c3f30ca3.chunk.js (common) 87.6 kB  [rendered]
chunk {1} 1.806b6d85871e191d087d.chunk.js () 334 kB  [rendered]
chunk {2} 2.84d2c2f503211e36c6b2.chunk.js () 633 kB  [rendered]
chunk {3} 3.8f203f13ef09c83a1cb4.chunk.js () 16.9 kB  [rendered]
chunk {4} 4.ee3dc3a237fdadde98e5.chunk.js () 38.2 kB  [rendered]
chunk {5} 5.c209b3bdcd81230bfdb7.chunk.js () 6.28 kB  [rendered]
chunk {6} 6.dd49bbc2f4e2cf3d3930.chunk.js () 72.1 kB  [rendered]
chunk {7} main.93622f1b31e94850762f.bundle.js (main) 2.69 MB [initial] [rendered]
chunk {8} polyfills.8f464027c5f775c4b960.bundle.js (polyfills) 66.5 kB [initial] [rendered]
chunk {9} styles.9487a01f65ac7a4f3242.bundle.css (styles) 179 kB [initial] [rendered]
chunk {10} inline.3d54bd4262d98168f168.bundle.js (inline) 1.63 kB [entry] [rendered]
```
With `comparisons: false`, prod build (mapbox-gl work):
```
chunk {0} 0.7120f44b0b35c3f30ca3.chunk.js (common) 87.6 kB  [rendered]
chunk {1} 1.806b6d85871e191d087d.chunk.js () 334 kB  [rendered]
chunk {2} 2.84d2c2f503211e36c6b2.chunk.js () 633 kB  [rendered]
chunk {3} 3.8f203f13ef09c83a1cb4.chunk.js () 16.9 kB  [rendered]
chunk {4} 4.ee3dc3a237fdadde98e5.chunk.js () 38.2 kB  [rendered]
chunk {5} 5.c209b3bdcd81230bfdb7.chunk.js () 6.28 kB  [rendered]
chunk {6} 6.dd49bbc2f4e2cf3d3930.chunk.js () 72.1 kB  [rendered]
chunk {7} main.93622f1b31e94850762f.bundle.js (main) 2.69 MB [initial] [rendered]
chunk {8} polyfills.8f464027c5f775c4b960.bundle.js (polyfills) 66.5 kB [initial] [rendered]
chunk {9} styles.9487a01f65ac7a4f3242.bundle.css (styles) 179 kB [initial] [rendered]
chunk {10} inline.3d54bd4262d98168f168.bundle.js (inline) 1.63 kB [entry] [rendered]
```
Yup, same size. I don't think there's any good reason to not disable `comparisons` at this point.

**Comment by SublimeProphets** [comment]
The PR works perfect for me.
I'm running angular 5.0.1 and angular-cli 1.5.0

**Comment by khanhvuNOIS** [comment]
The PR worked well

**Comment by filipesilva** [comment]
Heya, will be merging this into the next 1.6 release.

**Comment by kzc** [comment]
@filipesilva While I disagree with the [root cause](https://github.com/mishoo/UglifyJS2/issues/2520#issuecomment-347342508) of this issue,  if `mapbox-gl` will not address it then it would be better to disable the more specific uglify `compress` option `typeofs` rather than disabling `comparisons`.

There's a few optimizations unrelated to `mapbox-gl` that would be negatively affected with `comparisons` disabled.

Example code with default `compress` options:
```
$ echo 'function f(global){ var o = {}; console.log(o == o, typeof global === "undefined"); }' | bin/uglifyjs -c
function f(global){console.log(true,void 0===global)}
```
Same code with `comparisons` disabled:
```
$ echo 'function f(global){ var o = {}; console.log(o == o, typeof global === "undefined"); }' | bin/uglifyjs -c comparisons=false
function f(global){var o={};console.log(o==o,"undefined"===typeof global)}
```
And again the same code with `typeofs` disabled:
```
$ echo 'function f(global){ var o = {}; console.log(o == o, typeof global === "undefined"); }' | bin/uglifyjs -c typeofs=false
function f(global){console.log(true,"undefined"==typeof global)}
```
You'll notice that the contentious webpacked `mapbox-gl` `"undefined"==typeof global` is undisturbed while the other `comparisons` optimization is still performed.

Not a big deal in the grand scheme of things, but sometimes the `comparisons` transform can be used to eliminate unnecessary code.

File: packages/@angular/cli/models/webpack-configs/production.ts [hunk 1]
```diff
@@ -103,7 +103,13 @@ export function getProdConfig(wco: WebpackConfigOptions) {
     }));
   }
 
-  const uglifyCompressOptions: any = {};
+  const uglifyCompressOptions: any = {
+      // Disabled because of an issue with Uglify breaking seemingly valid code:
```
Comment (kzc) [inline packages/@angular/cli/models/webpack-configs/production.ts]: That statement is inaccurate. See https://github.com/mishoo/UglifyJS2/issues/2520#issuecomment-347342508

Comment (Wykks) [reply packages/@angular/cli/models/webpack-configs/production.ts]: I'm not too proficient in english, but it's written "an issue with Uglify", and not "an issue in Uglify". It's two different things, no ?

**Comment by filipesilva** [comment]
@kzc I was unaware that there was a better way to target the specific problem, and agree that it would be better. Would you be willing to put up a similar PR to makes it more specific?

For the record our intention to address this class of problems is to wholly disable the node `global` in webpack for browser apps in CLI 2.0. Merging this fix was an interim compromise to support `mapbox-gl` users.

File: packages/@angular/cli/models/webpack-configs/production.ts (hunk 1 above)
Comment (kzc) [reply packages/@angular/cli/models/webpack-configs/production.ts]: It's a mapbox-gl issue.

**Comment by kzc** [comment]
@filipesilva Since this project is one of the bigger users of `uglify-es`, I scan the issues occasionally and happened across this. I'll leave it to you whether or not to act on this information.

**Comment by filipesilva** [comment]
@kzc we really appreciate that of you, your insight is very useful for keeping our Uglify usage correct and bug free.

I put up a PR with your suggested improvement: https://github.com/angular/angular-cli/pull/8694

**Pull request merged**